### ISSUE TITLE
Inconsistent results between using vmap and using for loop on GPUs

### ISSUE THREAD

**Issue by Hwhitetooth** [google/jax#1816]
Hi JAX team,

I find some inconsistent results between using vmap and using for loop on GPUs. The following is an example to reproduce the error. You can also find the same code in this [colab](https://colab.research.google.com/drive/1U98lyA--8UmNGlan3u116I8H01mk6aWa).

```python
import jax
import jax.numpy as jnp

key = jax.random.PRNGKey(0)

key, subkey = jax.random.split(key)
W = jax.random.normal(subkey, (3, 3))

key, subkey = jax.random.split(key)
A = jax.random.normal(subkey, (1, 4, 3))

y1 = []
for i in range(A.shape[1]):
  ai = A[:, i]
  yi = jnp.dot(ai, W)
  y1.append(yi)
y1 = jnp.stack(y1, axis=1)

y2 = jax.vmap(jnp.dot, (1, None), 1)(A, W)

delta = y1 - y2
print(delta)
```
The output says
```
[[[ 0.0000000e+00  0.0000000e+00 -5.9604645e-08]
  [ 0.0000000e+00 -2.3841858e-07  2.3841858e-07]
  [ 2.3841858e-07  0.0000000e+00  0.0000000e+00]
  [ 1.8626451e-09  0.0000000e+00  0.0000000e+00]]]
```

Could you please take a look at it? I am looking forward to your response.

Thanks!

**Comment by shoyer** [comment]
These results like they are consistent within the limits of floating point accuracy for float32. In general JAX (along with all numerical computing libraries I'm familiar with) cannot guarantee that two mathematically equivalent ways of writing an expression will result in exactly the same result. This particularly true for highly optimized routines such as matrix multiplication, which use different code paths for different array shapes.
